### ISSUE TITLE
Run custom powershell script after the build step

### ISSUE THREAD

**Issue by PureKrome** [appveyor/ci#128]
Ref: http://help.appveyor.com/discussions/questions/860-how-can-i-run-my-own-powershell-script-after-the-build-step

Would be really nice to have the ability to execute a `powershell` script file, after the build step AND for a particular tag(s).

```
deploy:
- provider: Script
  script: do_something_here() <-- path to PS file
  on:
    branch: master
```

**Comment by danielmcormond** [comment]
Subscribed! We would definitely use this `Script` provider here at @RelayFoods if it becomes available.

**Comment by tadeumaia** [comment]
:+1:

**Comment by couellet** [comment]
That would be very useful!

**Comment by JakeGinnivan** [comment]
+1 would be super handy

**Comment by PureKrome** [comment]
Another reason why this is a +1 (even though I created this issue) is this scenario: -

Current AV project has a powershell script to do deployment. This is in the `on_success` stage. This will always deploy (when success) even though the build is a Pull Request.

Really need to be able to have a custom PS script in deployment :+1:

**Comment by JakeGinnivan** [comment]
My reasons are:

 - GitVersion needs to push a gem during deployment
 - Other private projects need to run DbUp against a database before deploying the web project

**Comment by PureKrome** [comment]
@JakeGinnivan DB migrations - of course! That's a great scenario! I didn't think of that..

**Comment by RamblingCookieMonster** [comment]
Another vote, would be helpful to have this abstracted into the yml, rather than in the script.

Cheers!

**Comment by jannickj** [comment]
What is the status on this, I've run into the problem and having deployment happen as part of pushing to master is just too ugly for my taste

**Comment by FeodorFitsner** [comment]
@jannickj could you please give more details what would be your use case for that?

**Comment by jannickj** [comment]
@FeodorFitsner we have a script that builds and deploys based on the offical f# setup for projects: https://github.com/fsprojects/ProjectScaffold
that uses this release process:
http://fsprojects.github.io/ProjectScaffold/release-process.html

Since the build logs are public it poses major security issues for the company.
And we have no way to hide our nuget and github keys from the build log.

**Comment by JakeGinnivan** [comment]
Id also love an update on this, I still cannot automate GitVersion deployments with AppVeyor. I really just want to do:

```
deploy_script:
  ps: appveyor create-variables releasenotes=releaseNotesGenerator
  ps: appveyor deploy-nuget ...
  ps: appveyor create-github-release .. /releaseNotes $releaseNotes$
  ps: gem publish ..
  ps: choco publish ...
```

Something along those lines would be amazing and would save me sooo much effort when doing a GitVersion release. It also would enable me to suggest AppVeyor to clients instead of TFS/Teamcity + octopus deploy.

**Comment by MikhailTymchukDX** [comment]
As far as I see, I can not run PowerShell script from a file yet?

**Comment by larsmathiesen** [comment]
It would be VERY helpful to be able to run a custom powershell script upon build success.

**Comment by MikhailTymchukDX** [comment]
@larsmathiesen As far as I understand, you can run it:
```
on_success:
- ps: '& "Folder\Script.ps1"'
```

**Comment by arekkas** [comment]
This is desperately needed, for example when using `electron-builder`!

**Comment by IlyaFinkelshteyn** [comment]
Sorry if I misunderstand the subject but it is always possible to verify if build is tagged in script. 

CMD example:
```
deploy_script:
- IF DEFINED APPVEYOR_REPO_TAG_NAME (build_script)
```

PowerShell example:
```
deploy_script:
- ps: IF ($env:APPVEYOR_REPO_TAG_NAME) {build_script}
```

Thanks!
Ilya.

**Comment by PureKrome** [comment]
@IlyaFinkelshteyn is the `deploy_script ` 

1. the full _replacement_ script/steps for custom deploying? This means the `deploy` settings are ignored.
2. a script that is ran _after_ the deploy has complete. This means the `deploy` settings are still used.

**Comment by IlyaFinkelshteyn** [comment]
@PureKrome 
 - `deploy_script` is full replacement for `deploy` settings
 - `after_deploy` can be used together with deploy settings to run *after* deploy is complete.

You can use `IF DEFINED APPVEYOR_REPO_TAG_NAME` condition in both of them.

I still feel that I don't understand the problem fully and there is some scenario which cannot be solved with those scripts and IFs...

**Comment by PureKrome** [comment]
Ok. lets create a fake scenario.

lets imagine .. i donno .. that after a build, we want to open the cage door to a carrier-pigeon which will then do something.

so we need to hit a url with a POST for example... `POST https://www.ci-pigeons.io/purekrome/project-69`

so can you write some yml for our deploy then.. like this..

```
deploy:
  - provider: WebDeploy
    server: https://<SNIP>.scm.azurewebsites.net:443/msdeploy.axd?site=<SNIP>
    website: <SNIP>
    username: $<SNIP>
    password:
      secure: <SNIP>
    artifact: Website.zip
    remove_files: true
    app_offline: true
    on:
      appveyor_repo_tag: false
```
(Deploy can be webdeploy, nuget deploy .. it's irrelevant... just _some_ Deploy section ... )

so somehow .. alter this `provider` settings to allow us to do some _post_ deploy powershell...

k-thx-go!

**Comment by PureKrome** [comment]
<deleted post> 

I was totally talking about the wrong things...

**Comment by PureKrome** [comment]
@IlyaFinkelshteyn i've totally re-read my original question and i think I've got it now.

Currently, there's a nice way to use the `Deploy` settings to setup up automatic deployment.
`- provider: WebDeploy`
`- provider: Ftp`
`- provider: etc...`

So what I was suggesting to @FeodorFitsner in the original question was to create _another_ type of provider: `- provider: Script` and this can be the location of a `ps1` file or even some embedded script.

Now, what you AV-krew are saying that, this is actually possible already!

Instead of doing:

```
deploy:
  - provider: Webdeploy
    ...
```

we should replace that section with: 
`deploy_script: <whatever>`

Yes / no ?

( i guess i couldn't find much info on `deploy_script` at the time when i originally asked the question).

----

@JakeGinnivan Jake - could you do the 100% deployment step in a custom powershell script?
So basically, can you do :

```
deploy_script:
  ps: appveyor create-variables releasenotes=releaseNotesGenerator
  ps: appveyor deploy-nuget ...
  ps: appveyor create-github-release .. /releaseNotes $releaseNotes$
  ps: gem publish ..
  ps: choco publish ...
```

Finally - i'm assuming the `deploy_script` is a valid setting that exists _now_ in appveyor.yml ...

**Comment by IlyaFinkelshteyn** [comment]
Need to clarify that this is still not the same as deployment provider. It is just single custom script which runs at deployment stage during a build. For example it cannot benefit from filtering syntax (like `on: branch: master`) and you need to do `IF`s in script itself.

It would be much better if you could create your own "custom script provider" and have it like this:
```
deploy:
  - provider: Script
    name: MyScript1
on:
    branch: master
```

Reason we still did not do it, is because providers should run in both **Inline** and **Environment** deployments. However we cannot allow to run arbitrary code in **Environment** deployments, because they run on central servers (not on build workers).

Good news is that this year we plan to implement new Docker-based deployment **Environments** which will allow to do custom scripts and hopefully we will be able to make "Script" deployment provider.

Sorry for confusion ;)

Ilya.

**Comment by PureKrome** [comment]
>It would be much better if you could create your own "custom script provider" and have it like this

yep - that's what i was suggesting:

>So what I was suggesting to @FeodorFitsner in the original question was to create another type of provider: - provider: Script and this can be the location of a ps1 file or even some embedded script.


---

Also @IlyaFinkelshteyn , can you answer the question about `deploy` vs `deploy_script`, plz? (refer ☝️ to previous reply).

**Comment by IlyaFinkelshteyn** [comment]
@PureKrome You understanding of `deploy_script` is correct (Yes on that question). 

Also you can experiment with UI -- go to `settings -> deployment -> script`, choose CMD or PowerShell, add some script code, save and press `Export YAML`. You will see sample syntax. However auto-generated (by `Export YAML`) multi-line PowerShell does not look very compact and if you don't like it, you can use [this samples](https://www.appveyor.com/docs/build-configuration/#script-blocks-in-build-configuration) instead.

Ilya.

**Comment by IlyaFinkelshteyn** [comment]
Closing in favor of https://github.com/appveyor/ci/issues/2042

**Comment by stephanmg** [comment]
Hello, if I use multiple `- ps:`  statements will variables be still set?
